src/storage-client/src/types/connections.rs

```diff
@@ -1029,3 +1056,200 @@ impl SshTunnel {
+use rdkafka::error::{KafkaError, KafkaResult};
+use rdkafka::producer::{DeliveryResult, ProducerContext};
+use rdkafka::types::RDKafkaRespErr;
+use rdkafka::util::Timeout;
+use rdkafka::{Statistics, TopicPartitionList};
+
+/// A client context that supports rewriting broker addresses,
+/// and is also able to spin up ssh tunnels dynamically.
+pub struct DynamicBrokerRewritingClientContext<C> {
+    pub inner: BrokerRewritingClientContext<C>,
+    default_override: Option<SshTunnel>,
+    secrets_reader: Arc<dyn SecretsReader>,
+    error: Mutex<Option<String>>,
+    runtime: Handle,
+}
+
+impl<C> DynamicBrokerRewritingClientContext<C> {
+    /// Constructs a new context that wraps `inner`.
+    pub fn new(
+        inner: C,
+        secrets_reader: Arc<dyn SecretsReader>,
+        runtime: Handle,
+    ) -> DynamicBrokerRewritingClientContext<C> {
+        DynamicBrokerRewritingClientContext {
+            inner: BrokerRewritingClientContext::new(inner),
+            default_override: None,
+            secrets_reader,
+            error: Mutex::new(None),
+            runtime,
+        }
+    }
+
+    /// Adds the default broker rewrite rule.
+    ///
+    /// Connections to brokers that aren't specified in other rewrites will be rewritten to connect to
+    /// `rewrite_host` and `rewrite_port` instead.
+    pub fn add_default_broker_rewrite(&mut self, tunnel: SshTunnel) {
+        self.default_override = Some(tunnel);
+    }
+
+    /// Returns a reference to the wrapped context.
+    pub fn inner(&self) -> &C {
+        self.inner.inner()
+    }
+
+    /// Returns any errors that occurred when setting up
+    /// ssh tunnels.
+    pub fn dynamic_error(&self) -> Option<String> {
+        self.error.lock().expect("poisoned").clone()
+    }
+}
+
+impl<C> ClientContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ClientContext,
+{
+    const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
+
+    fn rewrite_broker_addr(&self, addr: BrokerAddr) -> BrokerAddr {
+        if self.inner.overridden(&addr) {
+            return self.inner.rewrite_broker_addr(addr);
+        }
+
+        match &self.default_override {
+            Some(default_tunnel) => {
+                let broker = addr.clone();
+                let default_tunnel = default_tunnel.clone();
+                let sr = Arc::clone(&self.secrets_reader);
+                let runtime = self.runtime.clone();
+                // We spawn an additional thread here, and immediately
+                // join it so that the rdkafka thread calling this callback
+                // spends all its time parked, instead of being
+                // continuously woken up when we `block_on` the
+                // ssh tunnel creation.
+                let res = match std::thread::spawn(move || {
+                    runtime.block_on(async move {
+                        let ret = default_tunnel
+                            .build_ssh_tunnel_for_url(
```

Comment:
My kneejerk is that this would be a lot less gnarly if we split it up into two parts: a quick synchronous one that grabs a port, and a slower async one that stands up the ssh tunnel.

We could reasonably return a port and a token synchronously, without doing all the work of actually spinning up the ssh connection, and delegate that to some background task. It's possible that `rdkafka` would just spin for a bit trying to reconnect as the connection gets set up, but it does that all the time anyways. And if there are errors setting up the tunnel, we can report that elsewhere in the source pipeline... there's no particular reason it needs to be piped via the rdkafka context. (And frankly as you've noticed piping errors through the rdkafka machinery is not so nice, so I would consider this a plus.)